### ISSUE TITLE
incorrect error message: 'this' is not allowed before super

### ISSUE THREAD

**Issue by michaelficarra** [babel/babel#1131]
I am getting this error in code that does not use `super`.

```
SyntaxError: src/early-error-state.js: Line 77: 'this' is not allowed before super()
  75 | export class EarlyError extends Error {
  76 |   constructor(node, message) {
> 77 |     this.node = node;
     |     ^
  78 |     this.message = message;
  79 |   }
  80 | }
```

If that doesn't reproduce the problem on its own, I can create a reduced test case.

```
$ babel --version
5.0.1
```

**Comment by sebmck** [comment]
No this error is correct. You need to call `super()` in derived class constructors.

**Comment by michaelficarra** [comment]
I was not aware of that. Can you send me the relevant portion of the spec where that is enforced? Is it an early error or runtime error?

**Comment by sebmck** [comment]
`[[Construct]]` is not called in derived class constructors until a `SuperCall`:

> 7. Let _result_ be Construct(_func_, _argList_, _newTarget_).

http://people.mozilla.org/~jorendorff/es6-draft.html#sec-super-keyword-runtime-semantics-evaluation

Which AFAIK means that it has no `this` binding.

**Comment by sebmck** [comment]
FWIW, this is also the behaviour of Traceur and Typescript.

**Comment by michaelficarra** [comment]
> `[[Construct]]` is not called in derived class constructors until a `SuperCall`:

That is not true. `[[Construct]]` is called for anything that has `[[Construct]]` when `new` is used:
http://people.mozilla.org/~jorendorff/es6-draft.html#sec-evaluatenew

I think this behaviour is incorrect. At the very least, it should be a runtime error (though I'm quite confident it would not be), not a compile time error.

**Comment by sebmck** [comment]
I'll find a proper a proper reference in the morning. In the meantime if
you feel this behaviour is in fact incorrect then id recommend also opening
issues on Traceur and Typescript. I'm concerned about why you think it'd be
a good idea to delegate this to a runtime error as it's mean wrapping all
this expressions which would have an adverse performance cost.

On Friday, 3 April 2015, Michael Ficarra <notifications@github.com> wrote:

> [[Construct]] is not called in derived class constructors until a
> SuperCall:
>
> That is not true. [[Construct]] is called for anything that has
> [[Construct]] when new is used:
> http://people.mozilla.org/~jorendorff/es6-draft.html#sec-evaluatenew
>
> I think this behaviour is incorrect. At the very least, it should be a
> runtime error (though I'm quite confident it would not be), not a compile
> time error.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/babel/babel/issues/1131#issuecomment-89047374>.
>


-- 
Sebastian McKenzie

**Comment by michaelficarra** [comment]
A runtime error would simply be an unconditional throw in the marked position. It doesn't depend on a runtime value, so there's no need to wrap `this` references. Consider the following program, which will throw 50% of the time and succeed without error the other 50%.

```js
class A extends B {
  constructor() {
    if (Math.random() < 0.5) return
    this; // assuming this is an error in the absence of super, throws upon reaching this point
  }
}
```

I've never used the Traceur or Typescript projects, so I don't feel it'd be appropriate opening an issue on them, at least until the two of us can agree on which behaviour is correct.

**Comment by eventualbuddha** [comment]
Even weirder, what would the spec say about this?

```js
class A extends B {
  constructor() {
    if (Math.random() < 0.5) super();
    this;
  }
}
```

Babel seems happy enough to transpile it, but according to the spec it'll fail half the time (maybe).

**Comment by sebmck** [comment]
@michaelficarra Still need to call `super()` so I don't understand what you're getting at.

@eventualbuddha Yeah, you can't get it perfect and there are obviously a lot of scenarios you can't catch although it's good enough™ for 99% of users and is incredibly helpful.

**Comment by stefanpenner** [comment]
I believe this was in-fact a somewhat recent refinement to the spec. It boils down to `super` must be called before `this` is touched. This ensures the super hierarchy has sufficiently initialized the `this` before descendants have access.

In the `A < B` example, this ensures B has fully initialized the object before A can further refine it.

If needed I can gladly provide some more real life examples as to why this is a good thing.

**Comment by wycats** [comment]
The TLDR is:

In ES5, a JavaScript constructor is always initializing the same kind of allocated object: an Object. As a result, ES5 function constructors receive a newly allocated object at the beginning of the function, and can go ahead and initialize it.

In ES6, a JavaScript base class can allocate a special kind of object, like an Array, Date, RegExp, or a user-specified exotic object like a proxy. However, the decision about what kind of object to allocate is made in the base class, and until a subclass `super`s, the runtime doesn't know **what kind of object `this` is**.

```js
class Foo extends Bar {
  constructor() {
    // what is the internal representation of `this`? unknown
    super();
    // the internal representation of `this` is whatever the superclass provided
  }
}
```

This flexibility, which makes it possible to subclass Arrays, Dates and DOM classes, makes the meaning of `this` unknown until `super()` is invoked. That's a dynamic question that is answered by `super` returning, not a static one.

This requirement also ensures that a superclass constructor has had a chance to initialize itself before the allocated object is used. This means that if you write a base class, you don't have to worry about subclasses accidentally forgetting to call `super()` but still invoking methods that assume that initialization has taken place.

**Comment by eventualbuddha** [comment]
Good to know. Thanks for the explanation, @wycats.

**Comment by janmarek** [comment]
Is the error still correct when parent class does not have a constructor?

**Comment by sebmck** [comment]
Yes.

On Tuesday, 21 April 2015, Jan Marek <notifications@github.com> wrote:

> Is the error still correct when parent class does not have a constructor?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/babel/babel/issues/1131#issuecomment-94915822>.
>


-- 
Sebastian McKenzie

**Comment by hgl** [comment]
@wycats if `super()` always needs to be called before `this` is accessed, why not make `constructor()` call `super()` implicitly if this is a subclass? Forgetting to call `super()` seems will be a frequent mistake.

Maybe this topic should be discussed on es-discussion?

**Comment by hgl** [comment]
NVM, found the discussion: https://esdiscuss.org/topic/super-on-class-that-extends

**Comment by hgl** [comment]
and also realized `super()` can't be implicitly called because you might want to pass different arguments to it.

Sorry for the noises generated. :)

**Comment by chrisabrams** [comment]
What if I want to override the constructor? This means I cannot use `this` in that scenario?

**Comment by fkling** [comment]
@chrisabrams: You only have to call `super()` before you are accessing `this`. That's all.

**Comment by hayesmg** [comment]
I am attempting to pass a function to the superclass and still receiving this error.  Is it possible to have it not inspect _this_ when it is wrapped in a delegate?

```js
class Bar extends Foo {
  constructor() {
    super(() => this.doSomething());
  }

  doSomething() { }
}
```

**Comment by shama** [comment]
Just curious if anyone has a solution for this pattern being broken because of this:

```js
class Child extends Parent {
  constructor () {
    if (!(this instanceof Child)) return new Child();
    super();
  }
}
```

So the user can choose their preference when using the API:

```js
import Child from './child.js'
var child = new Child()

// OR

import createChild from './child.js'
var child = createChild()
```

**Comment by loganfsmyth** [comment]
@shama What you describe isn't really related to this issue. That pattern is not possible with ES6 classes anyway because the spec says that calling an ES6 class constructor without `new` throws an exception.

**Comment by shama** [comment]
Thanks @loganfsmyth.

I got an elegant workaround from @dlmanning:

```js
class Child extends Parent {
  constructor () {
    super();
  }
}
export default function ChildFactory (...args) {
  return new Child(...args);
}
```

Which gives me the API I'm looking for with `class`. Cheers!

**Comment by hgl** [comment]
@shama why do you hate `new` so much? :)

**Comment by shama** [comment]
@hgl I don't hate `new`. I want to build APIs that can cater to both preferences. That pattern is elegant because it doesn't require increasing the surface area of the API to accommodate both.

**Comment by hgl** [comment]
@shama Ah, I see your point. Useable with or without `new`.

**Comment by trusktr** [comment]
I dislike this new ES6 spec. Having logical control of when the super constructor is a good thing. It gives us more freedom of logic!

**Comment by trusktr** [comment]
How do we join the discussion on ES Discuss? I replied to the thread, but don't see my message. Do I have to wait for it to be approved?

**Comment by dlmanning** [comment]
@trusktr : Ship has sailed on this one, buddy. ES6 was finalized earlier this month. class syntax is a part of javascript now.

**Comment by jmm** [comment]
@trusktr I think you have to be subscribed first. Are you?

**Comment by loganfsmyth** [comment]
@trusktr At the end of the day, you either allow `this` before `super`, or allow subclassing `Error` and `Array`, but not both, and in the end I think creating custom errors and such is more important.

As others have said, we're well past being able to change this, it's really not worth arguing over at this point.

**Comment by sebmck** [comment]
I don't like locking issues but this issue has been discussed to death. It's pointless arguing this now and if you want to continue please do it on a medium that's not the Babel GitHub issue tracker.
